From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wea_ondara <wea_ondara@alpenblock.net>
Date: Fri, 18 Mar 2022 15:34:46 -0400
Subject: [PATCH] Performance improvement for Chunk.getEntities

This patch aims to reduce performance cost used by collecting the
entities of a chunk. Previously the entitySlices were copied into an
extra array with List.toArray() with is a costly and unneccessary
operation. This patch will reduce the load of plugins which for example
implement custom moblimits and depend on Chunk.getEntities().

diff --git a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
index 8a33e282ce9f4ead6828cd0323de29e2c02fb605..c2f902c15e27da820a2d7c0aaad6d32429e1c6ff 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
@@ -87,13 +87,15 @@ public class CraftChunk implements Chunk {
 
         for (int i = 0; i < 16; i++) {
 
-            for (Object obj : chunk.entitySlices[i].toArray()) {
-                if (!(obj instanceof net.minecraft.server.Entity)) {
+            // PandaSpigot start - speed up (was with chunk.entitySlices[i].toArray() and cast checks which costs a lot of performance if called often)
+            for (net.minecraft.server.Entity entity : chunk.entitySlices[i]) {
+                if (entity == null) {
                     continue;
                 }
 
-                entities[index++] = ((net.minecraft.server.Entity) obj).getBukkitEntity();
+                entities[index++] = entity.getBukkitEntity();
             }
+            // PandaSpigot end
         }
 
         return entities;
